File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/McpProtocol.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Text.Json
open System.Text.Json.Serialization

module McpProtocol =

    [<CLIMutable>]
    type JsonRpcRequest = {
        [<JsonPropertyName("jsonrpc")>]
        JsonRpc: string
        [<JsonPropertyName("id")>]
        Id: int64
        [<JsonPropertyName("method")>]
        Method: string
        [<JsonPropertyName("params")>]
        Params: obj
    }

    let createRequest (id: int64) (methodName: string) (paramsObj: obj) : string =
        let req = {
            JsonRpc = "2.0"
            Id = id
            Method = methodName
            Params = if isNull paramsObj then obj() else paramsObj
        }
        JsonSerializer.Serialize(req)

    let createInitializeRequest (id: int64) (clientName: string) (clientVersion: string) : string =
        let paramsObj = {|
            protocolVersion = "2024-11-05"
            capabilities = {| tools = {||} |}
            clientInfo = {|
                name = clientName
                version = clientVersion
            |}
        |}
        createRequest id "initialize" paramsObj

    let createToolsListRequest (id: int64) : string =
        createRequest id "tools/list" {| |}

    let createToolsCallRequest (id: int64) (toolName: string) (argumentsJson: string) : string =
        let parsedArgs =
            if String.IsNullOrWhiteSpace(argumentsJson) then
                JsonDocument.Parse("{}").RootElement
            else
                try JsonDocument.Parse(argumentsJson).RootElement
                with _ -> JsonDocument.Parse("{}").RootElement

        let paramsObj = {|
            name = toolName
            arguments = parsedArgs
        |}
        createRequest id "tools/call" paramsObj

File: implementations/csharp/src/Skight.AgentPlatform/McpProtocol.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Skight.AgentPlatform
{
    public class JsonRpcRequest
    {
        [JsonPropertyName("jsonrpc")]
        public string JsonRpc { get; set; } = "2.0";

        [JsonPropertyName("id")]
        public long Id { get; set; }

        [JsonPropertyName("method")]
        public string Method { get; set; } = string.Empty;

        [JsonPropertyName("params")]
        public object Params { get; set; } = new { };
    }

    public static class McpProtocol
    {
        public static string CreateRequest(long id, string methodName, object paramsObj)
        {
            var req = new JsonRpcRequest
            {
                JsonRpc = "2.0",
                Id = id,
                Method = methodName,
                Params = paramsObj ?? new { }
            };
            return JsonSerializer.Serialize(req);
        }

        public static string CreateInitializeRequest(long id, string clientName = "Skight.AgentPlatform", string clientVersion = "1.0.0")
        {
            var paramsObj = new
            {
                protocolVersion = "2024-11-05",
                capabilities = new { tools = new { } },
                clientInfo = new
                {
                    name = clientName,
                    version = clientVersion
                }
            };
            return CreateRequest(id, "initialize", paramsObj);
        }

        public static string CreateToolsListRequest(long id)
        {
            return CreateRequest(id, "tools/list", new { });
        }

        public static string CreateToolsCallRequest(long id, string toolName, string argumentsJson)
        {
            JsonElement parsedArgs;
            if (string.IsNullOrWhiteSpace(argumentsJson))
            {
                parsedArgs = JsonDocument.Parse("{}").RootElement;
            }
            else
            {
                try
                {
                    parsedArgs = JsonDocument.Parse(argumentsJson).RootElement;
                }
                catch
                {
                    parsedArgs = JsonDocument.Parse("{}").RootElement;
                }
            }

            var paramsObj = new
            {
                name = toolName,
                arguments = parsedArgs
            };
            return CreateRequest(id, "tools/call", paramsObj);
        }
    }
}
